contracts/Proxy.sol

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2020 Compound Labs, Inc.
+
+pragma solidity ^0.7.5;
+
+import "./ErrorReporter.sol";
+import "./ProxyAdminStorage.sol";
+
+/**
+ * @dev This abstract contract provides a fallback function that delegates all calls to another
+ * contract using the EVM instruction `delegatecall`. We refer to the second contract as the
+ * _implementation_ behind the proxy.
+ *
+ * Additionally, delegation to the implementation can be triggered manually through the {_fallback}
+ * function, or to a different contract through the {_delegate} function.
+ *
+ * The success and return data of the delegated call will be returned back to the caller of the
+ * proxy.
+ */
+contract Proxy is ProxyAdminStorage, ErrorReporter {
+    /**
+     * @notice Emitted when pendingImplementation is changed
+     */
+    event NewPendingImplementation(
+        address oldPendingImplementation,
+        address newPendingImplementation
+    );
+
+    /**
+     * @notice Emitted when pendingImplementation is accepted, which means comptroller
+     * implementation is updated
+     */
+    event NewImplementation(address oldImplementation, address newImplementation);
+
+    /**
+     * @notice Emitted when pendingAdmin is changed
+     */
+    event NewPendingAdmin(address oldPendingAdmin, address newPendingAdmin);
+
+    /**
+     * @notice Emitted when pendingAdmin is accepted, which means admin is updated
+     */
+    event NewAdmin(address oldAdmin, address newAdmin);
+
+    constructor() {
+        // Set admin to caller
+        admin = msg.sender;
+    }
+
+    /*** Admin Functions ***/
+    function _setPendingImplementation(address newPendingImplementation) public returns (uint256) {
+        if (msg.sender != admin) {
+            return fail(Error.UNAUTHORIZED, FailureInfo.SET_PENDING_IMPLEMENTATION_OWNER_CHECK);
+        }
+
+        address oldPendingImplementation = pendingImplementation;
+
+        pendingImplementation = newPendingImplementation;
+
+        emit NewPendingImplementation(oldPendingImplementation, pendingImplementation);
+
+        return uint256(Error.NO_ERROR);
+    }
+
+    /**
+     * @notice Accepts new implementation. `msg.sender` must be pendingImplementation
+     * @dev Admin function for new implementation to accept its role as implementation
+     * @return uint 0=success, otherwise a failure (see ErrorReporter.sol for details)
+     */
+    function _acceptImplementation() public returns (uint256) {
+        // Check caller is pendingImplementation and pendingImplementation ≠ address(0)
+        if (msg.sender != pendingImplementation || pendingImplementation == address(0)) {
+            return
+                fail(Error.UNAUTHORIZED, FailureInfo.ACCEPT_PENDING_IMPLEMENTATION_ADDRESS_CHECK);
+        }
+
+        // Save current values for inclusion in log
+        address oldImplementation = implementation;
+        address oldPendingImplementation = pendingImplementation;
+
+        implementation = pendingImplementation;
```

Comment:
Shouldn't we call `pendingImplementation.upgrade(self)`? Otherwise it doesn't have a right to veto the upgrade and a chance to perform any migrations.

Reply:
The upgrade process is started from this contract (`_setPendingImplementation`), but completed from the contract being upgraded. So `_acceptImplementation` is called by the target contract, after it has prepared for an upgrade. See here: https://github.com/radicle-dev/radicle-contracts/pull/73/files#diff-822ef81b3cb234d98d0343ec37a053b9c6a62a021b027e87e2f4e8cb646180d6R10